spopt/locate/p_median.py

```diff
@@ -0,0 +1,233 @@
+        )
+        FacilityModelBuilder.add_opening_constraint(
+            p_median, p_median.problem, r_fac, r_cli
+        )
+
+        return p_median
+
+    @classmethod
+    def from_geodataframe(
+        cls,
+        gdf_demand: GeoDataFrame,
+        gdf_fac: GeoDataFrame,
+        demand_col: str,
+        facility_col: str,
+        weights_cols: str,
+        p_facilities: int,
+        distance_metric: str = "euclidean",
+        name: str = "p-median",
+    ):
+        """
+        Create a PMedian object based on geodataframes. Calculate the cost matrix between demand and facility,
+        and then use from_cost_matrix method.
+
+        Parameters
+        ----------
+        gdf_demand: geopandas.GeoDataFrame
+            demand geodataframe with point geometry
+        gdf_fac: geopandas.GeoDataframe
+            facility geodataframe with point geometry
+        demand_col: str
+            demand geometry column name
+        facility_col: str
+            facility candidate sites geometry column name
+        weights_cols: str
+            weight column name representing service load or demand
+        p_facilities: int
+            number of facilities to be located
+        distance_metric: str, default="euclidean"
+            metrics supported by :method: `scipy.spatial.distance.cdist` used for the distance calculations
+        name: str, default="p-median"
+            name of the problem
+
+        Returns
+        -------
+        PMedian object
+        """
+
+        service_load = gdf_demand[weights_cols].to_numpy()
+        dem = gdf_demand[demand_col]
+        fac = gdf_fac[facility_col]
+
+        dem_type_geom = dem.geom_type.unique()
+        fac_type_geom = fac.geom_type.unique()
+
+        if len(dem_type_geom) > 1 or not "Point" in dem_type_geom:
+            warnings.warn(
+                "Demand geodataframe contains mixed type geometries or is not a point. Be sure deriving centroid from geometries doesn't affect the results.",
+                Warning,
+            )
+            dem = dem.centroid
+
+        if len(fac_type_geom) > 1 or not "Point" in fac_type_geom:
+            warnings.warn(
+                "Facility geodataframe contains mixed type geometries or is not a point. Be sure deriving centroid from geometries doesn't affect the results.",
+                Warning,
+            )
+            fac = fac.centroid
+
+        dem_data = np.array([dem.x.to_numpy(), dem.y.to_numpy()]).T
+        fac_data = np.array([fac.x.to_numpy(), fac.y.to_numpy()]).T
+
+        distances = np.array([])
+
+        if gdf_demand.crs != gdf_fac.crs:
+            raise ValueError(
+                f"geodataframes crs are different: gdf_demand-{gdf_demand.crs}, gdf_fac-{gdf_fac.crs}"
+            )
+
+        distances = cdist(dem_data, fac_data, distance_metric)
+
+        return cls.from_cost_matrix(distances, service_load, p_facilities, name)
+
+    def facility_client_array(self) -> None:
```

Comment:
Is there no tests for `facility_client_array()`? It seems lines 204-217 are not being hit during CI. 

*This also applies to the `PCenter`, `LSCP`, and `MCLP` classes.*